File: crates/mathhook-python/examples/advanced_usage.py
#!/usr/bin/env python3
"""
MathHook Python Advanced Usage Examples

This example demonstrates advanced mathematical operations,
complex expressions, and real-world use cases for the
MathHook Python bindings.

Last Updated: 2025-12-28T1200
"""

import time
from mathhook import (
    symbol, symbols, parse, solve,
    sin, cos, tan, exp, log, sqrt, abs_expr,
    factorial, gamma,
    pprint, init_printing,
    gcd, lcm,
    PyODESolver, PyExpression
)

def main():
    print("=" * 70)
    print("MathHook Python - Advanced Usage Examples")
    print("=" * 70)

    # =========================================================================
    # Complex Mathematical Expressions
    # =========================================================================
    print("\n1. Complex Mathematical Expressions")
    print("-" * 50)

    x, y, z = symbols('x y z')
    a, b, c = symbols('a b c')

    # Multi-variable polynomial: ax² + bxy + cy²
    multi_poly = a * x**2 + b * x * y + c * y**2
    print(f"Multi-variable polynomial: {multi_poly}")
    print(f"Simplified: {multi_poly.simplify()}")

    # Nested operations: (x + y)³
    cubed = (x + y) ** 3
    print(f"\n(x + y)³ = {cubed}")
    print(f"Expanded: {cubed.expand()}")

    # Complex fraction: (x² + 2x + 1) / (x + 1)
    numerator = x**2 + 2*x + 1
    denominator = x + 1
    fraction = numerator / denominator
    print(f"\n(x² + 2x + 1) / (x + 1) = {fraction}")
    print(f"Simplified: {fraction.simplify()}")

    # =========================================================================
    # Calculus Operations
    # =========================================================================
    print("\n2. Calculus Operations")
    print("-" * 50)

    # Derivative
    expr = x**3 + 2*x**2 + x + 1
    print(f"f(x) = {expr}")
    derivative = expr.derivative('x')
    print(f"f'(x) = {derivative}")

    # Second derivative
    second_derivative = expr.nth_derivative('x', 2)
    print(f"f''(x) = {second_derivative}")

    # Integration
    integral = expr.integrate('x')
    print(f"∫f(x)dx = {integral}")

    # Trigonometric calculus
    trig_expr = sin(x) * cos(x)
    print(f"\ng(x) = sin(x)cos(x)")
    print(f"g'(x) = {trig_expr.derivative('x')}")

    # Chain rule: d/dx[sin(x²)]
    chain_expr = sin(x**2)
    print(f"\nh(x) = sin(x²)")
    print(f"h'(x) = {chain_expr.derivative('x')}")

    # Product rule: d/dx[x²sin(x)]
    product_expr = x**2 * sin(x)
    print(f"\np(x) = x²sin(x)")
    print(f"p'(x) = {product_expr.derivative('x')}")

    # =========================================================================
    # Equation Solving
    # =========================================================================
    print("\n3. Equation Solving")
    print("-" * 50)

    # Linear equation: 2x + 3 = 7
    solutions = solve(2*x + 3 - 7, 'x')
    print(f"2x + 3 = 7 → x = {solutions}")

    # Quadratic equation: x² - 5x + 6 = 0
    solutions = solve(x**2 - 5*x + 6, 'x')
    print(f"x² - 5x + 6 = 0 → x = {solutions}")

    # Quadratic with complex roots: x² + 1 = 0
    solutions = solve(x**2 + 1, 'x')
    print(f"x² + 1 = 0 → x = {solutions}")

    # Cubic equation: x³ - 6x² + 11x - 6 = 0
    solutions = solve(x**3 - 6*x**2 + 11*x - 6, 'x')
    print(f"x³ - 6x² + 11x - 6 = 0 → x = {solutions}")

    # =========================================================================
    # Substitution
    # =========================================================================
    print("\n4. Substitution")
    print("-" * 50)

    expr = x**2 + 2*x + 1
    print(f"f(x) = {expr}")

    # Substitute x = 3
    result = expr.substitute({'x': PyExpression.integer(3)})
    print(f"f(3) = {result}")

    # Substitute x = y + 1
    result = expr.substitute({'x': y + 1})
    print(f"f(y + 1) = {result}")
    print(f"f(y + 1) expanded = {result.expand()}")

    # =========================================================================
    # Limits
    # =========================================================================
    print("\n5. Limits")
    print("-" * 50)

    # Limit of sin(x)/x as x → 0
    expr = sin(x) / x
    limit_result = expr.limit('x', PyExpression.integer(0))
    print(f"lim(x→0) sin(x)/x = {limit_result}")

    # Limit of (x² - 1)/(x - 1) as x → 1
    expr = (x**2 - 1) / (x - 1)
    limit_result = expr.limit('x', PyExpression.integer(1))
    print(f"lim(x→1) (x² - 1)/(x - 1) = {limit_result}")

    # =========================================================================
    # Polynomial Operations
    # =========================================================================
    print("\n6. Polynomial Operations")
    print("-" * 50)

    p1 = x**2 - 1
    p2 = x - 1

    # GCD
    result = gcd(p1, p2)
    print(f"gcd(x² - 1, x - 1) = {result}")

    # Multiplication
    product = p1 * p2
    print(f"(x² - 1)(x - 1) = {product.expand()}")

    # Division
    quotient = p1 / p2
    print(f"(x² - 1)/(x - 1) = {quotient.simplify()}")

    # =========================================================================
    # Trigonometric Identities
    # =========================================================================
    print("\n7. Trigonometric Identities")
    print("-" * 50)

    # Pythagorean identity: sin²(x) + cos²(x) = 1
    identity = sin(x)**2 + cos(x)**2
    print(f"sin²(x) + cos²(x) = {identity.simplify()}")

    # Double angle: 2sin(x)cos(x) = sin(2x)
    double_angle = 2 * sin(x) * cos(x)
    print(f"2sin(x)cos(x) = {double_angle.simplify()}")

    # =========================================================================
    # Logarithmic Operations
    # =========================================================================
    print("\n8. Logarithmic Operations")
    print("-" * 50)

    # Log product rule: log(xy) = log(x) + log(y)
    log_sum = log(x) + log(y)
    print(f"log(x) + log(y) = {log_sum.simplify()}")

    # Log power rule: log(x^n) = n*log(x)
    log_power = log(x**3)
    print(f"log(x³) = {log_power.simplify()}")

    # =========================================================================
    # Special Functions
    # =========================================================================
    print("\n9. Special Functions")
    print("-" * 50)

    n = symbol('n')

    print(f"factorial(5) = {factorial(5)}")
    print(f"gamma(5) = {gamma(5)}")
    print(f"exp(log(x)) = {exp(log(x)).simplify()}")

    # =========================================================================
    # ODE Solving (Numerical)
    # =========================================================================
    print("\n10. ODE Solving (Numerical)")
    print("-" * 50)

    # Solve dy/dx = -2y with y(0) = 1
    t = symbol('t')
    y_var = symbol('y')

    # Using PyODESolver for numerical ODE solving
    ode_solver = PyODESolver()

    print("ODE: dy/dt = -2y, y(0) = 1")
    print("Analytical solution: y(t) = e^(-2t)")

    # The numerical solver would be used for more complex ODEs
    # For simple exponential decay, we can verify:
    analytical = exp(-2 * t)
    result = analytical.substitute({'t': PyExpression.integer(1)})
    print(f"At t=1: y(1) = {result}")

    # =========================================================================
    # Performance Demo
    # =========================================================================
    print("\n11. Performance Demo")
    print("-" * 50)

    start = time.time()
    expressions = []
    for i in range(1000):
        expr = x * i + y**2 + i * 2
        expressions.append(expr.simplify())
    elapsed = (time.time() - start) * 1000

    print(f"Created and simplified 1000 expressions in {elapsed:.2f}ms")
    print(f"Average: {elapsed/1000:.4f}ms per expression")

    # =========================================================================
    # Summary
    # =========================================================================
    print("\n" + "=" * 70)
    print("Advanced usage examples completed!")
    print("=" * 70)


if __name__ == "__main__":
    main()

File: crates/mathhook-node/examples/advanced-usage.ts
#!/usr/bin/env tsx

/**
 * Advanced MathHook Node.js/TypeScript Usage Examples
 *
 * This example demonstrates advanced mathematical operations,
 * complex expressions, and real-world use cases.
 */

import { Expression, symbols, parse, SmartEquationSolver, sin, cos, exp } from "../index.js";

console.log("MathHook TypeScript Advanced Usage Examples\n");

// ===== Complex Mathematical Expressions =====
console.log("Complex Mathematical Expressions:");

// Variables for advanced examples
const [x, y, z, a, b, c] = symbols("x y z a b c");

// Multi-variable polynomial: ax^2 + bxy + cy^2
const multiPoly = Expression.add([
  Expression.mul([a, Expression.pow(x, Expression.integer(2))]),
  Expression.mul([b, x, y]),
  Expression.mul([c, Expression.pow(y, Expression.integer(2))]),
]);

console.log(`Multi-variable polynomial: ${multiPoly.format()}`);
console.log(`Simplified: ${multiPoly.simplify().format()}\n`);

// ===== Nested Operations =====
console.log("Nested Mathematical Operations:");

// Nested expression: (x + y)^3
const cubed = Expression.pow(Expression.add([x, y]), Expression.integer(3));
console.log(`(x + y)^3: ${cubed.format()}`);
console.log(`Simplified: ${cubed.simplify().format()}`);

// Complex fraction-like expression: (x^2 + 2x + 1) / (x + 1)
const numerator = Expression.add([
  Expression.pow(x, Expression.integer(2)),
  Expression.mul([Expression.integer(2), x]),
  Expression.integer(1),
]);
const denominator = Expression.add([x, Expression.integer(1)]);
const fraction = Expression.div(numerator, denominator);

console.log(`Complex fraction: ${fraction.format()}`);
console.log(`Simplified: ${fraction.simplify().format()}\n`);

// ===== System of Equations (Conceptual) =====
console.log("System of Equations Example:");

try {
  // First equation: 2x + 3y = 12
  const eq1Left = Expression.add([
    Expression.mul([Expression.integer(2), x]),
    Expression.mul([Expression.integer(3), y]),
  ]);
  const eq1 = Expression.equation(eq1Left, Expression.integer(12));

  // Second equation: x - y = 1
  const eq2Left = Expression.add([x, Expression.mul([Expression.integer(-1), y])]);
  const eq2 = Expression.equation(eq2Left, Expression.integer(1));

  console.log(`Equation 1: ${eq1.format()}`);
  console.log(`Equation 2: ${eq2.format()}`);

  // Note: Current solver handles single equations
  // System solving would require additional implementation
  console.log("Note: System solving requires additional implementation\n");
} catch (error: unknown) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`System solver error: ${message}\n`);
}

// ===== Advanced Parsing Examples =====
console.log("Advanced Parsing Examples:");

try {
  // Complex LaTeX expressions - automatic detection
  const latexExpressions = [
    "\\sqrt{x^2 + y^2}",
    "\\frac{a^2 + b^2}{c^2}",
    "x^{2n+1} + y^{n-1}",
  ];

  console.log("LaTeX Expressions (Auto-detected):");
  latexExpressions.forEach((expr, index) => {
    try {
      const parsed = parse(expr);
      console.log(`  ${index + 1}. "${expr}" => ${parsed.format()}`);
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : String(e);
      console.log(`  ${index + 1}. "${expr}" => Parse Error: ${message}`);
    }
  });

  // Simple mathematical expressions
  const simpleExpressions = [
    "x^2 + 2*x + 1",
    "sin(x) + cos(y)",
    "exp(-x^2/2)",
  ];

  console.log("\nSimple Expressions:");
  simpleExpressions.forEach((expr, index) => {
    try {
      const parsed = parse(expr);
      console.log(`  ${index + 1}. "${expr}" => ${parsed.format()}`);
    } catch (e: unknown) {
      const message = e instanceof Error ? e.message : String(e);
      console.log(`  ${index + 1}. "${expr}" => Parse Error: ${message}`);
    }
  });
} catch (error: unknown) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`Advanced parser error: ${message}`);
}

console.log("\n");

// ===== Performance Testing =====
console.log("Performance Testing:");

const performanceTest = () => {
  const startTime = process.hrtime.bigint();

  // Create and simplify 1000 expressions
  for (let i = 0; i < 1000; i++) {
    const expr = Expression.add([
      Expression.mul([x, Expression.integer(i)]),
      Expression.pow(y, Expression.integer(2)),
      Expression.integer(i * 2),
    ]);
    expr.simplify();
  }

  const endTime = process.hrtime.bigint();
  const duration = Number(endTime - startTime) / 1_000_000; // Convert to milliseconds

  console.log(
    `Created and simplified 1000 expressions in ${duration.toFixed(2)}ms`
  );
  console.log(`Average: ${(duration / 1000).toFixed(4)}ms per expression`);
};

performanceTest();

// ===== Memory Usage Example =====
console.log("\nMemory Usage Example:");

const memoryTest = () => {
  const expressions: Expression[] = [];

  // Create a large number of expressions
  for (let i = 0; i < 10000; i++) {
    const expr = Expression.add([
      Expression.pow(x, Expression.integer(i % 5)),
      Expression.mul([y, Expression.integer(i)]),
      Expression.pow(z, Expression.integer(2)),
    ]);
    expressions.push(expr);
  }

  console.log(`Created ${expressions.length} expressions in memory`);

  // Simplify all expressions
  const simplified = expressions.map((expr) => expr.simplify());
  console.log(`Simplified ${simplified.length} expressions`);

  // Clear references (JavaScript GC will handle cleanup)
  expressions.length = 0;
  simplified.length = 0;

  console.log("Memory test completed");
};

memoryTest();

// ===== Real-world Use Case: Quadratic Formula =====
console.log("\nReal-world Use Case: Quadratic Formula");

const quadraticFormula = () => {
  try {
    const solver = new SmartEquationSolver();

    // Quadratic equation: ax^2 + bx + c = 0
    // For example: x^2 - 5x + 6 = 0
    const a_val = Expression.integer(1);
    const b_val = Expression.integer(-5);
    const c_val = Expression.integer(6);

    const quadratic = Expression.add([
      Expression.mul([a_val, Expression.pow(x, Expression.integer(2))]),
      Expression.mul([b_val, x]),
      c_val,
    ]);

    const equation = Expression.equation(quadratic, Expression.integer(0));

    console.log(`Quadratic equation: ${equation.format()}`);

    const [solution, explanation] = solver.solveWithEquation(equation, x.asSymbol()!);
    console.log(`Solution valid: ${solution.isValidSolution()}, count: ${solution.solutionCount()}`);

    // The solutions should be x = 2 and x = 3 (since (x-2)(x-3) = x^2 - 5x + 6)
  } catch (error: unknown) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Quadratic formula error: ${message}`);
  }
};

quadraticFormula();

// ===== Trigonometric Expression Building =====
console.log("\nTrigonometric Expression Building:");

try {
  // sin^2(x) + cos^2(x)
  const trig = Expression.add([
    Expression.pow(sin(x), Expression.integer(2)),
    Expression.pow(cos(x), Expression.integer(2)),
  ]);
  console.log(`sin^2(x) + cos^2(x) = ${trig.format()}`);
  console.log(`Simplified: ${trig.simplify().format()}`);

  // exp(i*x) relationship
  const complex = exp(Expression.mul([Expression.i(), x]));
  console.log(`exp(i*x) = ${complex.format()}`);
} catch (error: unknown) {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`Trig error: ${message}`);
}

console.log("\nAll advanced examples completed successfully!");
console.log(
  "You now know how to use MathHook with TypeScript for complex mathematical operations!"
);
